### PR TITLE
Handle redirects in book scan

### PR DESCRIPTION
**Pull request by jarosik10**
for: https://github.com/openstax/unified/issues/1556

### CODE REVIEW

File: script/utils/bookUtils.ts
Comment (PiotrKozlowski) [inline script/utils/bookUtils.ts:18]: loaders could be next to each other and rootUrl may be still first parameter
they don't have to be in alphabetic order and it will create less diff

File: script/domVisitor.ts
Comment (PiotrKozlowski) [inline script/domVisitor.ts:84]: You could change the `checkIfPageHasRedirect` a little bit and do something like this:
```suggestion
      const validatedUrl = getPageUrlOrRedirection(redirects, pageUrl);
      await page.goto(`${rootUrl}${validatedUrl}${appendQueryString}`);
```

**Comment by jarosik10** [comment]
@TomWoodward @PiotrKozlowski 
To be honest I'm not sure how test properly.

File: script/urlChecker.ts
Comment (PiotrKozlowski) [inline script/urlChecker.ts:57]: I think there is no need to first go to the page and then go to the redirected page in case of 404
You should do the same as in `domVisitior` where you first check for the redirected url

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]

File: script/utils/getPageUrlOrRedirection.ts
Comment (PiotrKozlowski) [inline script/utils/getPageUrlOrRedirection.ts:4]: it is returning `Redirect | string`, but should always return `string`
Also, I think you should compare `page` to `redirect.to` and return `redirect.from` in order to check if it is working because it is checking on `openstax.org` that doesn't have a new version yet

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]
You can test it by merging https://github.com/openstax/rex-web/pull/1165/ to your branch and running `validate-modified-books.bash` script.

You can see that it is failing on one page: https://github.com/openstax/rex-web/pull/1165/checks?check_run_id=2481010263

You'll need to update `getPageUrlOrRedirection.ts`

File: script/urlChecker.ts
Comment (TomWoodward) [inline script/urlChecker.ts:47]: this is gonna return _one of_ the redirects targeting the current page in no particular order, i think we need to check all of them

File: script/urlChecker.ts
Comment (TomWoodward) [inline script/urlChecker.ts:49]: the criteria should be
1) if the page url returns a 200 on prod, 👍 
2) if the page url is a redirect target, one of the redirects returns a 200 on prod, 👍

File: script/urlChecker.ts
Comment (TomWoodward) [inline script/urlChecker.ts:98]: idk if there is any point in running this

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: script/urlChecker.ts
Comment (jarosik10) [reply script/urlChecker.ts:98]: @TomWoodward 
What do you mean? Do I have access to `rex/redirects.json` during CI?

Comment (TomWoodward) [reply script/urlChecker.ts:98]: `prepareRedirects` doesn't add any information you don't already have, it just renders the book id and page id into a url, you're comparing the pages to urls and comparing the urls, you could just compare the ids

**Comment by jarosik10** [comment]
There must be a issue with `status === 200`?

File: script/urlChecker.ts
Comment (TomWoodward) [inline script/urlChecker.ts:104]: what is the point of reformatting this array?

**Review by TomWoodward — CHANGES_REQUESTED** [review]

File: script/urlChecker.ts
Comment (PiotrKozlowski) [inline script/urlChecker.ts:96]: you can return an empty array here and you will not need and `if` statement in `visitPage`

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]

File: script/urlChecker.ts
Comment (PiotrKozlowski) [inline script/urlChecker.ts:53]: You don't have to use `map`, you can destructure below

**Comment by jarosik10** [comment]
Reminder: undo commit https://github.com/openstax/rex-web/pull/1170/commits/fc2f223b61ed1d9e6cbadabff026e6a229a5ada1 before merging.

File: script/urlChecker.ts
Comment (PiotrKozlowski) [inline script/urlChecker.ts:52]: you could change the name of this variable and unshift it with `pageURL` so you can do everything in one for loop

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]
It looks good. I've left 2 small comments.

We have to remember about removing updates to `/data/redirects/..` and `config.books.json`

File: script/urlChecker.ts
Comment (jarosik10) [reply script/urlChecker.ts:52]: I'd have to add `pageURL` as `urls.unshift({pathname: pageURL} as any)`. Not sure if it is okay to have such a weird array. Leaving `map` and then adding `unshift` will create array of strings, which seems to be more clear.

Comment (PiotrKozlowski) [reply script/urlChecker.ts:52]: right, in this case, it makes sense to use `map`
I actually wrote these comments in another order :D

File: script/urlChecker.ts
Comment (PiotrKozlowski) [inline script/urlChecker.ts:59]: you probably want to `return false` after the loop to keep types consistent

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]
+ https://github.com/openstax/rex-web/pull/1170#issuecomment-839740016

**Review by TomWoodward — COMMENTED** [review]
looks good, need to remove that debug commit before approval

**Review by PiotrKozlowski — APPROVED** [review]

**Review by TomWoodward — APPROVED** [review]

**Pull request merged**
